### PR TITLE
Stop document context manager from saving if an error is raised.

### PR DESCRIPTION
**Pull request by smithsz**
## Checklist

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)
- [x] Added tests for code changes _or_ test/build only changes
- [x] Updated the change log file (`CHANGES.md`|`CHANGELOG.md`) _or_ test/build only changes
- [x] Completed the PR template below:

## Description

Fix bug where document CM performed save in event of uncaught exception.

## Schema & API Changes

No change.

## Security and Privacy

No change.

## Testing

Includes 2 additional unit tests.

## Monitoring and Logging

No change.

### CODE REVIEW

File: src/cloudant/document.py [hunk 1]
```diff
@@ -341,12 +341,13 @@ def __enter__(self):
 
         return self
 
-    def __exit__(self, *args):
+    def __exit__(self, exc_type, exc_value, traceback):
         """
-        Support context like editing of document fields.  Handles context exit
-        logic.  Executes a Document.save() upon exit.
+        Support context like editing of document fields. Handles context exit
+        logic. Executes a `Document.save()` upon exit iff no exception occurred.
```
Comment (emlaver) [inline src/cloudant/document.py]: iff -> if

File: tests/unit/document_tests.py [hunk 2]
```diff
@@ -622,6 +622,47 @@ def test_document_context_manager_no_doc_id(self):
...
+        doc['age'] = 6
+        doc.save()
+
+        # Make a document update and then raise an error.
+        with self.assertRaises(ZeroDivisionError), Document(self.db, 'julia006') as doc:
+            doc['age'] = 7
+            raise ZeroDivisionError()
+
+        # Assert the change persists locally.
+        self.assertEqual(doc['age'], 7)
```
Comment (smithsz) [inline tests/unit/document_tests.py]: Note, if we raise an exception in the block then the local copy still gets updated. It's only the remote save that gets skipped.

File: src/cloudant/document.py (hunk 1 above)
Comment (smithsz) [reply src/cloudant/document.py]: I was abbreviating _if and only if_ but I guess plain old "if" works and will cause less confusion.

File: tests/unit/document_tests.py (hunk 2 above)
Comment (emlaver) [reply tests/unit/document_tests.py]: I wonder if there's a way for us to use a temporary dict during the context manager that only updates the local doc and save remotely if there's no exception.
If not, we should add a note in the context manager docs that if an exception occurs, the remote doc is not saved but the local doc is.

Comment (smithsz) [reply tests/unit/document_tests.py]: I'll add a further note in the docs. I think a temp dict could add too many complications.

File: docs/getting_started.rst [hunk 3]
```diff
@@ -441,41 +441,45 @@ multiple updates to a single document. Note that we don't save to the server
 after each update. We only save once to the server upon exiting the ``Document``
 context manager.
 
- .. code-block:: python
+.. warning:: Uncaught exceptions inside the ``with`` block will prevent your
```
Comment (smithsz) [inline docs/getting_started.rst]: I've just added this warning. The other changes in this file are for correctly indenting the following code blocks.

File: CHANGES.md [hunk 4]
```diff
@@ -1,7 +1,8 @@
 # Unreleased
 
-- [IMPROVED] Updated `Getting started` section with a `get_query_result` example. 
+- [FIXED] Bug where document CM performed save in event of uncaught exception.
```
Comment (emlaver) [inline CHANGES.md]: We should spell out context manager

File: docs/conf.py [hunk 5]
```diff
@@ -21,6 +21,8 @@
 # documentation root, use os.path.abspath to make it absolute, like shown here.
 #sys.path.insert(0, os.path.abspath('.'))
 
+sys.path.insert(0, os.path.abspath('../src'))
```
Comment (emlaver) [inline docs/conf.py]: I've had to use this in the past to get documentation builds to properly run.  I don't think it's needed for building in `readthedocs.org`.

Comment (smithsz) [reply docs/conf.py]: Yeah. I always have to add it in too. I'll leave it and hopefully it'll make building locally easier.

**Pull request merged**